src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILImporter.Interpreter.cs

```diff
@@ -159,9 +175,9 @@ private void ImportLoadNull()
         private void ImportReturn()
         {
             var returnType = _method.Signature.ReturnType;
-            if (returnType.IsVoid)
+            if (returnType.RuntimeTypeHandle.Value == typeof(void).TypeHandle.Value)
```

Comment:
What was wrong with `IsVoid` ? The new code looks less efficient.

Reply:
There was a bug that caused `returnType.IsVoid` to return false when it was supposed to be true. I can't exactly remember the issue but it seems to have been fixed now on the latest master so will revert

Reply:
There's a bug that causes `returnType.IsVoid` to return false even when the method is actually void. 
This line https://github.com/dotnet/corert/blob/master/src/Common/src/TypeSystem/Common/TypeDesc.cs#L149 evaluates to `ValueType` instead of `Void`. I'm guessing the code building the type description sets `_typeFlags` to `ValueType` because a Void is defined as a `struct`. I hit this issue really early on so it skipped my mind to report it

Reply:
I have added the following assert to right after the type system context gets created and it always passes fine:

```
TypeDesc voidType = context.ResolveRuntimeTypeHandle(typeof(void).TypeHandle);
Debug.Assert(voidType.IsVoid);
```

Reply:
Adding the following at the bottom of the `ComputeTypeFlags` method (just before we return `flags`) fails the program:

```csharp
if (_typeName == "Void")
{
    Debug.Assert((flags & TypeFlags.Void) != 0);
}
```

Stepping through to https://github.com/dotnet/corert/blob/89f6a3e5c3e0d8a80031c3150a5360ffe9189797/src/Common/src/TypeSystem/NativeFormat/NativeFormatType.cs#L304 and I see that `_metadataUnit.Context.SystemModule` returns `null` instead of `System.Private.CoreLib` that `_module` is set to, causing the `TryGetCategoryFlagsForPrimitiveType` method to return `false` and default to assigning the `ValueType` flag.

Reply:
Can you try changing this line to use some other type (e.g. `bool`)?

https://github.com/dotnet/corert/blob/6b6cfd0a10ccf282144757555ae4a2116d289b0e/src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/TypeLoaderTypeSystemContext.cs#L39

I think we are racing between setting the system module and creating type flags for void.

If it fixes the IsVoid issue, we'll need to think of some other way to find the system module.

Reply:
Actually, switching this to WellKnownType.Object might be just what we need here. It doesn't have any special type flags.

Reply:
Thank you! Setting it to `Object` fixed it!

Reply:
Out of curiosity, when you say "racing" do you mean "race condition"? I didn't get the sense that I was dealing with multiple threads when I was stepping through the code

Reply:
No, I just meant initialization order; picked the wrong word - sorry for confusion. Things can get multithreaded in the type loader, but there should be no threads here.